bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/BidiUtil.java

```diff
@@ -456,66 +456,78 @@ public static void removeLanguageListener (long hwnd) {
 	removeLanguageListener(control.handle);
 }
+
 /**
  * Determine the base direction for the given text. The direction is derived
- * from that of the first strong bidirectional character. In case the text
- * doesn't contain any strong characters, the base direction is to be
- * derived from a higher-level protocol (e.g. the widget orientation).
+ * from the first strong bidirectional RIGHT_TO_LEFT character. Or if that does
+ * not exist from the first strong LEFT_TO_RIGHT character
  * <p>
  *
- * @param text
- *            Text base direction should be resolved for.
+ * @param text Text base direction should be resolved for.
  * @return SWT#LEFT_RIGHT or SWT#RIGHT_TO_LEFT if the text contains strong
  *         characters and thus the direction can be resolved, SWT#NONE
  *         otherwise.
  * @since 3.105
  */
-public static int resolveTextDirection (String text) {
-	if (text == null) return SWT.NONE;
-	int length = text.length();
-	if (length == 0) return SWT.NONE;
-	char[] rtlProbe = {' ', ' ', '1'};
-	char[] ltrProbe = {'\u202b', 'a', ' '};
-	char[] numberProbe = {'\u05d0', ' ', ' '};
-	GCP_RESULTS result = new GCP_RESULTS();
-	result.lStructSize = GCP_RESULTS.sizeof;
-	int nGlyphs = result.nGlyphs = ltrProbe.length;
-	long hHeap = OS.GetProcessHeap();
-	long lpOrder = result.lpOrder = OS.HeapAlloc(hHeap, OS.HEAP_ZERO_MEMORY, nGlyphs * 4);
-	long hdc = OS.GetDC(0);
-	int[] order = new int[1];
+public static int resolveTextDirection(String text) {
+	if (text == null)
+		return SWT.NONE;
 	int textDirection = SWT.NONE;
-	for (int i = 0; i < length; i++) {
-		char ch = text.charAt(i);
-		rtlProbe[0] = ch;
-		OS.GetCharacterPlacement(hdc, rtlProbe, rtlProbe.length, 0, result, OS.GCP_REORDER);
-		OS.MoveMemory(order, result.lpOrder, 4);
-		if (order[0] == 2) {
-			textDirection = SWT.RIGHT_TO_LEFT;
-			break;
-		}
-		if (textDirection == SWT.LEFT_TO_RIGHT) {
-			// If textDirection is already LTR, skip probing for LTR again.
-			continue;
+	for (int i = 0; i < text.length(); i++) {
+		char c = text.charAt(i);
+		byte directionality = Character.getDirectionality(c);
+		int strongDirection = getStrongDirection(directionality);
+		if (strongDirection != SWT.NONE) {
+			textDirection = strongDirection;
 		}
-		ltrProbe[2] = ch;
-		OS.GetCharacterPlacement(hdc, ltrProbe, ltrProbe.length, 0, result, OS.GCP_REORDER);
-		OS.MoveMemory(order, result.lpOrder + 4, 4);
-		if (order[0] == 1) {
-			numberProbe[2] = ch;
-			OS.GetCharacterPlacement(hdc, numberProbe, numberProbe.length, 0, result, OS.GCP_REORDER);
-			OS.MoveMemory(order, result.lpOrder, 4);
-			if (order[0] == 0) {
-				textDirection = SWT.LEFT_TO_RIGHT;
-				// Do-not break here, instead scan the complete text for any RTL possibility
-			}
+		if (textDirection == SWT.RIGHT_TO_LEFT) {
+			break;
 		}
 	}
-	OS.ReleaseDC (0, hdc);
-	OS.HeapFree(hHeap, 0, lpOrder);
 	return textDirection;
+}
+
+public static int getStrongDirection(byte directionality) {
```

Comment:
This method should be non-public, default access should be fine:
static int getStrongDirection(byte directionality)